library/core/src/ptr/mod.rs

```diff
@@ -582,12 +582,13 @@ pub const fn invalid_mut<T>(addr: usize) -> *mut T {
 /// Convert an address back to a pointer, picking up a previously 'exposed' provenance.
 ///
 /// This is equivalent to `addr as *const T`. The provenance of the returned pointer is that of *any*
-/// pointer that was previously passed to [`expose_addr`][pointer::expose_addr] or a `ptr as usize`
-/// cast. If there is no previously 'exposed' provenance that justifies the way this pointer will be
-/// used, the program has undefined behavior. Note that there is no algorithm that decides which
-/// provenance will be used. You can think of this as "guessing" the right provenance, and the guess
-/// will be "maximally in your favor", in the sense that if there is any way to avoid undefined
-/// behavior, then that is the guess that will be taken.
+/// pointer that was previously exposed by passing it to [`expose_addr`][pointer::expose_addr],
+/// or a `ptr as usize` cast. In addition, memory which is outside of the control of Rust
```

Comment:
Is "outside the control of Rust" sufficiently precise? Basically the important point is that it is disjoint from any Rust allocations (be they static, stack, heap, whatever).

Reply:
I agree that this language is quite vague and probably not precise enough. Is it possible for us to enumerate all the kinds of memory that must be explicitly exposed? For example...

If I know there is memory at some point defined by the hardware, and configure it as my stack, is all stack memory then exposed? I would expect the functions I execute in this context to be optimized normally, but surely it is legal `from_exposed_addr` any address which is the address of a stack allocation. But also, this memory is definitely not distinct from any Rust allocations. (this scenario is not too far away from what `loom` does, it uses `generator` to call `mmap` with `MAP_STACK` and set up the stack as the mapped memory)

Reply:
If your answer is yes, it is illegal if there is collision with a stack allocation, what about an `Allocator` implementation which hands out allocations to the memory? At which point does this "memory defined by the hardware" become heap? Does that `Allocator` just not allocate heap memory? What if I plug it into `#[global_allocator]`?

Reply:
When you set up some memory as stack memory, or return it from `global_allocator`, you transfer ownership of it to the Rust AM, so it is no longer "external"/"outside the AM".

Reply:
Hmm, unless it's previously been exposed, right?

Reply:
When memory becomes stack memory, it is 'de-exposed'. When it is exposed again, then yes it can be accessed via future from_exposed_addr again.

Reply:
What do you mean by "set up memory as stack memory"? If I'm for some reason programming without physical memory, as soon as Rust begins executing, is every address in the direction of stack growth transferred to the AM?

So to be clear does `Allocator` for sure not have the ownership-changing property? Or is it unclear?

Reply:
> What do you mean by "set up memory as stack memory"? If I'm for some reason programming without virtual memory (or actually... even if I do have virtual memory I guess this applies? How would LLVM know the difference), as soon as Rust begins executing, is every address in the direction of stack growth transferred to the AM?

Every address until the guard page. If there is no guard page or the target does not support stack probing... well then in principle we're already dead as we cannot prevent the stack from running over the edge of the address space.

I don't know how one sets up memory as stack memory, and that is certainly extremely target-dependent, but above it sounded like you were assuming such a mechanism when you said "configure it as my stack".

> So to be clear does Allocator for sure not have the ownership-changing property? Or is it unclear?

Hm... not sure if I like making that magic, but for now it's better not to commit either way.

Reply:
> in principle we're already dead as we cannot prevent the stack from running over the edge of the address space.

I don't understand this argument. It's perfectly possible to write programs that have a static cap on the amount of stack space they occupy.

My concern is that if the AM claims all memory in the direction of stack growth, that we will rule out using Rust on platforms where there are important outside-the-AM addresses in the direction of stack growth, but which the programmer ensures their stack never reaches. I would strongly prefer that this by a dynamic requirement: that the UB occurs only if a stack allocation actually collides with an outside-the-AM address. Of course this could have some odd properties like removing UB in the presence of optimizations, but the way we enforce limits on stack growth normally are not considered by the optimizer, so if this has to be a static requirement I don't understand how your point about "in principle we're already dead" doesn't always apply.

Reply:
> I don't understand this argument. It's perfectly possible to write programs that have a static cap on the amount of stack space they occupy.

rustc makes no guarantees on that front whatsoever.  So you can't really write any code where the SAFETY comment is "I know the stack will not grow larger than 1MIB" -- there is no principled basis for that assumption.

Pragmatically, of course, you can do a lot.

But anyway, we don't need to answer that question for this PR, do we? My comment was just asking whether we should amend "outside the control of Rust" by saying that in particular, this memory must be disjoint from stack, heap, and static memory, plus any other administrative memory the AM might use for itself (such as vtables).

Reply:
What do you mean by (such as vtables)? Aren't vtables stored in on the stack or heap?

Reply:
> So you can't really write any code where the SAFETY comment is "I know the stack will not grow larger than 1MIB" -- there is no principled basis for that assumption.

I think you could write something like:
```rs
// Safety: Our releases must undergo static assembly analysis that ensure
// the compiled code cannot use more than $size kB of stack memory.
```

E.g. you can ensure that this upholds after the code is compiled by inspecting the assembly. This kind of thing is not unheard of, FWIW. I've used https://github.com/HBehrens/puncover/ to do it, although that tool doesn't support Rust.

Reply:
See https://github.com/rust-lang/unsafe-code-guidelines/issues/338 -- vtables are magic and not technically memory, at least for now.

> you can ensure that this upholds after the code is compiled by inspecting the assembly

Yeah that sounds like a plausible strategy.

Reply:
@RalfJung I think the current change in this PR achieves the goal here, per the PR description. Do you agree? At the moment I think it would be better to not mention vtables, I don't think that adds any clarity.